Fix #18509: Adds the translation editing feature in the exploration editor page

DubeySandeep opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->
1. This PR fixes #18509
2. This PR does the following: Adds the translation editing feature in the exploration editor page
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

[scrnli_28_06_2023_04-00-18.webm](https://github.com/oppia/oppia/assets/16653571/ca8a753f-2e1e-4e6b-ae62-184bf8e27bee)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

DubeySandeep committed:
Adds translation editing feature in exploration editor page

DubeySandeep committed:
Merge branch 'develop' of github.com:oppia/oppia into translation-editor

DubeySandeep committed:
Handle change list in order

DubeySandeep committed:
Fix test issues

DubeySandeep committed:
Fix bacned tests

DubeySandeep committed:
Fix tests

sagangwee commented on core/domain/translation_services.py:
```diff
@@ -125,41 +125,98 @@ def add_new_translation(
+    entity_translation: translation_domain.EntityTranslation,
+    translation_changes: List[exp_domain.ExplorationChange]
+) -> None:
+    """Applies the changes to the entity_translations object.
```
```suggestion
    """Applies the changes to the entity_translation object.
```

DubeySandeep replied:
Done!

sagangwee commented on core/domain/translation_services.py:
```diff
@@ -125,41 +125,98 @@ def add_new_translation(
+                change
+            )
+
+            if entity_translation.language_code != change.language_code:
```
Should we validate all the changes upfront and throw an error if any of them have the incorrect language code? It seems a bit unexpected for some of the changes to silently fail.

DubeySandeep replied:
We only validate the schema of change list and internal values are validated inside `validation` method, in this case it will be validated after the entityTranslation.validate method is called. (This is similar to writtenTranslation and voiceover filed of exploration state.)

DubeySandeep replied:
I'm fine adding the validation here if you want it to be added here!

sagangwee replied:
Yes, that would be great if you don't mind. The method contract would be more explicit and easier to reason about in the future if we had to debug this.

sagangwee commented on core/domain/translation_services_test.py:
```diff
@@ -289,6 +296,139 @@ def test_compute_translation_related_change_mark_translation_needs_update(
+            )
+        )
+
+        entity_translation_models: Sequence[
+            translation_models.EntityTranslationsModel
+        ] = translation_models.EntityTranslationsModel.get_all().fetch()
+        self.assertEqual(len(entity_translation_models), 1)
+        entity_translation_model = entity_translation_models[0]
+        self.assertEqual(entity_translation_model.entity_version, 5)
+        self.assertEqual([
+            t['needs_update']
+            for t in entity_translation_model.translations.values()
+        ], [False, False])
```
Nit: remove the new lines above and below to consolidate the test setup phase or add comments to describe each "section". Ditto below.

DubeySandeep replied:
This is not setup it's the baseline verification. This test has the following sections separated by newlines: 1. Base setup 2. Basline verification 3. Setup for test 4. Action 5. Final verififcation

DubeySandeep replied:
I've added a new line between "4. Action 5. Final verification"

sagangwee commented on core/domain/translation_services_test.py:
```diff
@@ -289,6 +296,139 @@ def test_compute_translation_related_change_mark_translation_needs_update(
+            translation_services.compute_translation_related_change(
+                self.exp, change_list
+            )
+        )
```
Nit: Add new line to differentiate between the action under test and the main assertion. Ditto below.

DubeySandeep replied:
Done!

sagangwee reviewed (changes requested):
@DubeySandeep Thanks, left a few comments on codeowner files.

DubeySandeep committed:
Address review comments

DubeySandeep commented:
@sagangwee  @kevintab95 Can you please review this PR? (We are trying to get this in the next release as it's important for updating existing translations and somehow blocking the Android release.)

sagangwee commented on core/domain/translation_services_test.py:
```diff
@@ -289,6 +299,140 @@ def test_compute_translation_related_change_mark_translation_needs_update(
+            'ar' in [et.language_code for et in entity_translation_models]
+        )
+
+    def test_invalid_translation_change_raise_error(self) -> None:
```
Tiny nit:

```suggestion
    def test_invalid_translation_change_raises_error(self) -> None:
```

sagangwee reviewed (approved):
@DubeySandeep LGTM for codeowners minus a couple of minor comments. Thanks!

kevintab95 reviewed (commented):
LGTM but one question. Given that this is a feature, do we need to gate it behind a flag?

DubeySandeep commented:
@kevintab95 I think this s not a new feature and is more around re-adding the existing feature back. I'm not sure whether this should go with the feature flag. I'm happy to add this with a  feature flag, @seanlip do you think this need to be gated via a feature flag? (Can be helpful in case we want to disable it in the future)

seanlip commented:
Yes, this fixes a regression (the editing functionality was there previously). So I don't think we need a feature flag for it.

seanlip commented:
@DubeySandeep There were some CI failures -- I restarted, but if they happen again, please could you take a look? Thanks!

kevintab95 commented:
Thanks @DubeySandeep for clarifying. Approved!

seanlip commented:
@DubeySandeep could you please fix the CI issues?

DubeySandeep commented:
@seanlip I'll fix the tests tonight!

kevintab95 commented:
This will need to be hotfixed as the release is going out in the next 24 hours.

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 committed:
Set entity translation before each test

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 committed:
Mock local storage to get the correct language code

Existing code modifies the function globally which interferes with
other tests.

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 committed:
Improve coverage: TranslatedContentObjectFactory

kapilag96 committed:
Improve coverage: ChangeListService

Also fix misnamed action ("remove_translation" -> "remove_translations")

kapilag96 committed:
Improve coverage: EntityTranslationObjectFactory

Also rename the test file to *.spec.ts.

kapilag96 committed:
Add a test to check No callback on MarkAudioAsNeedingUpdateModalComponent

kapilag96 committed:
Remove export

kapilag96 committed:
Add a test

kapilag96 committed:
Add a test

kapilag96 committed:
Remove empty line

kapilag96 committed:
Fix ts errors

kapilag96 committed:
Undo data format changes

kapilag96 committed:
Improve coverage: state-translation-editorwq

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 commented:
Frontend tests and coverage has been fixed. Only thing left now is typescript compilation errors

DubeySandeep committed:
Fix typescript issues

kapilag96 committed:
Merge branch 'develop' into translation-editor

kapilag96 committed:
Merge branch 'develop' into translation-editor
